aesara/graph/opt.py

```diff
@@ -638,6 +638,38 @@ def process_node(self, fgraph, node):
                     # They were already tried, and there was an error
                     continue
 
+                # Check if outputs of node are all super types of the candidate outputs
+                if not all(
+                    node_out.type.is_super(candidate_out.type)
+                    for node_out, candidate_out in zip(node.outputs, candidate.outputs)
```

Comment:
Given how often merging is done, we should be careful about performance in these areas, so let's see if we can reformulate this condition and the following one to involve only a single pass through the candidates.

Reply:
I didn't spot a/the place where we can control the order over which we process nodes. In any  case I am not too happy with this and would rather just let #915 take care of it.

Reply:
Couldn't both directions be computed in one pass over `zip(node.outputs, candidate.outputs)`?

For example:
```python
                type_orders_1, type_orders_2 = zip(
                    *(
                        (
                            node_out.type.is_super(candidate_out.type),
                            candidate_out.type.is_super(node_out.type),
                        )
                        for node_out, candidate_out in zip(
                            node.outputs, candidate.outputs
                        )
                    )
                )
                # Check if outputs of node are all super types of the candidate outputs
                if not all(type_orders_1):
                    # If not, check if reverse is true
                    if not all(type_orders_2):
                        continue
```

Reply:
Ahh I thought you mean't the recalling process_node logic. 

I actually don't know how often we have this situation of mergeable nodes with different output types. I assumed it was rather low since it only showed in one very specific PyMC test so far, so that's why I kept the checks separately

Also this way it short circuits with `all`

Reply:
First off, I don't know if those two distinct `all` conditions are even necessary.  It's possible that one of the `is_super` calls can be short-circuited (e.g. using `or` between the two) in a single pass through the pairs that also subsumes the two `if` conditions.

Reply:
I need to know which of the two is True, one is to proceed as usual, the other is to reverse things. In any case I am much more worried about the performance penalty of recalling `process_node` than the second loop of type comparisons.

Reply:
A quick benchmark suggests the zip option to be slower:

```python
import aesara.tensor as at

def separate_loop(vars1, vars2):
    type_orders1 = (
        var1.type.is_super(var2.type)
        for var1, var2 in zip(vars1, vars2)
    )
    type_orders2 = (
        var2.type.is_super(var1.type)
        for var2, var1 in zip(vars2, vars1)
    )
    return all(type_orders1) or all(type_orders2)

def symultaneous_loop(vars1, vars2):
    type_orders_1, type_orders_2 = zip(
        *(
            (
                var1.type.is_super(var2.type),
                var2.type.is_super(var1.type),
            )
            for var1, var2 in zip(vars1, vars2)
        )
    )

    return all(type_orders_1) or all(type_orders_2)


type1 = at.TensorType("float64", shape=(1, 2, None, 3))
type2 = at.TensorType("float64", shape=(1, 2, None, None))
vars1 = (type1(), type1(), type1())
vars2 = (type2(), type2(), type2())

%timeit separate_loop(vars1, vars2)
%timeit separate_loop(vars2, vars1)
%timeit symultaneous_loop(vars1, vars2)
%timeit symultaneous_loop(vars2, vars1)
4.59 µs ± 31.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
3.51 µs ± 6.79 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.42 µs ± 130 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.36 µs ± 32.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
I don't see how the zip case can ever be faster from those results. Also I don't quite get what is the performance concern with the second separate loop.

Reply:
I did try different numbers and orders this was actually one of the more comparable cases

Reply:
> I don't see how the zip case can ever be faster from those results. Also I don't quite get what is the performance concern with the second separate loop.

The `zip` _isn't_ faster, and it wasn't supposed to be; it was meant to illustrate that both conditions can be handled with a single pass.

The approach I'm talking about simplifies the two conditions using a single pass over the pairs.  The example I gave doesn't actually do that, though.  It only illustrates how both conditions are related via a single pass.  The next step involves some consideration of what `is_super` and those two conditions are attempting to do.

Reply:
Recall that `TensorType.is_super` is supposed to be a partial ordering.  More specifically, it serves the purpose of `>=` between types, which is something like a product order.  (I believe the docstring for `Type.is_super` says `>`, so those are out of sync.)

With that in mind, it looks like your first condition checks that not all `a_n >= b_n` (i.e. that at least one instance of `a_n < b_n` is true) across all `n`, and the second condition checks that not all `b_n >= a_n` are true (i.e. that at least one instance of `b_n < a_n` is true).

In this setting, the general assumption is that the `Type`s are generally equal except for differences in their static shape specificity, and we can reduce the problem to a few simple cases:
1. all of the `candidate` outputs' static shapes are equal or more specific than `node`'s,
2. all of the `candidate` outputs' static shapes are less specific than `node`'s,
3. or some mix of the two.

The implicit `else` branch of the first `if` condition appears to handle 1. (i.e. queue the replacement), the implicit `else` branch of the second `if` handles 2. (i.e. attempt to queue the reverse replacement), and the first branch of the second `if` handles 3. (i.e. ignore the replacement).

The formulation of these three cases makes it seem like a second `is_super` pass is necessary, as well as the more costly recursive calls to `process_node`; however, the basic idea is that we want to replace less specific `Type`s with more specific&mdash;yet equivalent&mdash;ones (i.e. when `a <= b` use `Type` `a` instead of `b`).  

Without these case constraints we could apply the `<=`/`is_super` ordering across each pair of corresponding outputs (only _once_) and pick the "smallest"/most specific `Type` in each case.  If we could do this, then we would also avoid discarding candidates at this stage (which we never did previously) and the aforementioned performance concerns.

The apparent challenge here is that the most informative set of output `Type`s for the corresponding `Op` and inputs might not already exist as an `Apply` node in the graph (i.e. it isn't `node` or `candidate`); however, nothing is preventing us from creating and using such an `Apply` node.

In other words, if we track signatures comprised of `Op`, inputs, and outputs (with the "smallest" `Type`s) tuples&mdash;instead of effectively just the first two&mdash;we could consistently narrow the `Type`s in a graph and only perform a single `min`-based pass over node outputs and avoid recursion.